### ISSUE TITLE
[Feature request] Make it possible to show volumes on the desktop

### ISSUE THREAD

**Issue by probonopd** [lxqt/pcmanfm-qt#1192]
##### Expected Behavior

There should be a way to show volumes/partitions on the `--desktop` view, like they already show up under "Devices" in the side panel of the the file manager.

Most desktops allow this. E.g., Xfce:

![image](https://user-images.githubusercontent.com/2480569/92158067-a1932180-ee2b-11ea-9f43-d3ebae826168.png)

##### Current Behavior

Volumes/partitions cannot be shown in the `--desktop` view.

**Comment by probonopd** [comment]
There could be an option to enable this alongside the existing options:

![image](https://user-images.githubusercontent.com/2480569/92158274-f171e880-ee2b-11ea-94c4-4a646580161e.png)

**Comment by probonopd** [comment]
Unlike the existing options, what I am suggesting would __not__ just create some `.desktop` files. Instead, it would add icons on the desktop as volumes/partitions come and go, without creating `.desktop` files for those.

**Comment by probonopd** [comment]
Looks like in `desktopwindowmodel.c` one would need to use `modelFromPath(fm_path_new_for_uri("computer://"))` somehow in addition to what is already there, so that the "normal" icons _and_ the ones for `computer://` would get shown. What would be the best way to achieve this?

**Comment by tsujan** [comment]
Sorry for the late reply -- I was so busy.

First, this is a valid feature request.

Second, it seems that you were talking about a very old version of pcmanfm-qt or, perhaps, about the abandoned GTK version. We don't have `desktopwindowmodel.c`.

Third, the side-pane is completely different from `Fm::FolderView`. In the former, we deal with items; in the latter, with files and only files. That's the reason why Home, Trash, Computer,... are desktop entries.

Generalizing `Fm::FolderModelItem` to include non-files was my first thought before adding those items as desktop entries but it proved to be so complex and, especially, so dirty that I abandoned it.

If you can do it by writing a clean and concise code, your PR will be welcome.

**Comment by probonopd** [comment]
>you were talking about a very old version of pcmanfm-qt

Indeed, I am currently experiementing with my fork of 0.10.0 but I think the general gist of the ticket should still apply to the latest one.

**Comment by tsujan** [comment]
> if I change the model that is used for the desktop,

If you take a glance at the latest code (or the code from last year or 2 years ago), you'll see that we don't use a different model for desktop. The original author simplified the code and it should remain that way.

Anyhow, in this case too (like in https://github.com/lxqt/pcmanfm-qt/issues/1193), lets' forget about words. A clean and relatively short code will be acceptable for this.

**Comment by coldReactive** [comment]
Also would like to see this possible with udiskie and udisks2 support maybe, to get away completely from gnome dependencies with lxqt.

**Comment by tsujan** [comment]
>  ... to get away completely from gnome dependencies with lxqt.

LXQt has no GNOME dependency. GNOME even doesn't have a desktop, in the true sense of the word "desktop".

**Comment by coldReactive** [comment]
> > ... to get away completely from gnome dependencies with lxqt.
> 
> LXQt has no GNOME dependency. GNOME even doesn't have a desktop, in the true sense of the word "desktop".

gvfs will, however, get some gnome dependencies, such as gsettings-desktop-schemas

**Comment by probonopd** [comment]
Things like libfm, gvfs are arguably deep in "Gnome land".

**Comment by tsujan** [comment]
Using a fundamental library developed by the GNOME team doesn't mean having a GNOME dependency. GNOME is a DE based on GTK, while pcmanfm-qt and libfm-qt are Qt programs that use GLib for file management because GLib has proven to be superior to its Qt/KDE counterpart(s). GVFS extends GLib's functionalities a lot.

Anyway, that has nothing to do with this feature request.

**Comment by ahsand97** [comment]
@tsujan Since I got myself already involved in this haha I want to understand the current picture and ofcourse and if this would be worth it.

Right now:
```
DesktopWindow (the desktop's QListView)
        │
        ▼
Fm::ProxyFolderModel   (sorting/filtering)
        │
        ▼
Fm::FolderModel   (the actual list of rows)
        │
        ▼
each row = FolderModelItem, and every single one
           is required to wrap a real Fm::FileInfo
           (an actual file/.desktop entry on disk)
```

Volumes aren't files, so today there's just no slot for them anywhere in this stack.

So to solve this I was thinking in 2 approaches, one already mentioned by you.

## First approach: generalize FolderModelItem

```
DesktopWindow (QListView)
        │
        ▼
Fm::ProxyFolderModel        ← untouched
        │
        ▼
Fm::FolderModel              ← untouched class
        │
        ▼
   FolderModelItem  ──┬── file-backed   (Home, Trash, regular files — as today)
                      └── volume-backed  (NEW — a USB stick, a DVD, etc.)
```
but you said yourself it proved to be complex and dirty so I don't even know if I should try, maybe my implementation is "different" and it would worth a PR since you said also you're open to a PR.

## Second approach: keep two separate models, glue them together
Don't touch `FolderModel` at all. Build a second, brand-new, totally separate model just for volumes, and stitch the two together using a Qt tool built for exactly this `QConcatenateTablesProxyModel`:

```
                    DesktopWindow (QListView)
                             │
                             ▼
              QConcatenateTablesProxyModel   ← NEW glue layer
                   /                    \
                  ▼                      ▼
    Fm::ProxyFolderModel          Fm::VolumeListModel   ← brand new
           │                              │
           ▼                              ▼
    Fm::FolderModel                Fm::VolumeManager
    (real files only,               (USB drives, DVDs — 
     completely unchanged)            already exists, just unused)
```

The thing is, the second one requires more code than the first one, why? there's a big caveat:

```
        User clicks/selects a row in the QListView
                          │
                          ▼
             QModelIndex arrives at Fm::FolderView
                          │
              ┌───────────┴────────────┐
              │                        │
         Path A                    Path B
    "translate, then reuse      "just ask the row
     old code unchanged"          what it is"
              │                        │
              ▼                        ▼
   index → mapToSource()      index.data(FolderModel::FileInfoRole)
   → get model_'s own         → returns the file if it's a row,
     native index back          returns nothing/invalid if it's a
   → keep calling               volume row
     fileInfoFromIndex()       → this is the SAME pattern
     exactly like today          onItemActivated() and
                                  indexForPos() already
                                  use safely today
              │                        │
              ▼                        ▼
          needs Qt 6.11           works today, on 6.6.0 (minimum required Qt version currently)
        (mapToSource() isn't        no version bump needed
         available at the
         project's floor)        but it does mean rewriting
                                  selectedFilePaths(),
                                  selectedFiles(), selectFiles(),
                                  invertSelection(),
                                  childDragMoveEvent(),
                                  childDropEvent(),
                                  indexFromFolderPath()
                                  to use this pattern instead
                                  of the current unguarded
                                  model_->fileInfoFromIndex()
```

**Comment by tsujan** [comment]
Both may be possible but an overkill for a duplicate feature.

> Since I got myself already involved in this

I may succeed in involving you in something better: https://github.com/lxqt/libfm-qt/issues/417

**Comment by ahsand97** [comment]
> Both may be possible but an overkill for a duplicate feature.

I know but I still want this to be implemented (I'm very persistent haha). bumping Qt to 6.11 would be possible? if that's the case I could give a try to approach 2, because honestly I don't know if it'll be good to edit all the mentioned methods to not use `model_->fileInfoFromIndex()`.

Current logic:

```c++
auto file = model_->fileInfoFromIndex(*it);
paths.push_back(file->path());   // crash if *it isn't one of model_'s own rows
```

Proposed logic (without bumping Qt's version):
```c++
if(index.model()) {
    QVariant data = index.model()->data(index, FolderModel::FileInfoRole);
    if(data.isValid()) {
        auto file = data.value<std::shared_ptr<const Fm::FileInfo>>();
        if(file) {
            // ... use file ...
        }
    }
}
```

Bumping Qt's version:
```c++
QModelIndex sourceIndex = concatModel->mapToSource(index);   // ← the blocked method (Qt 6.11+)

if(sourceIndex.model() == proxyModel_) {
    // it came from the real-files half — safe to treat it as before
    auto file = proxyModel_->fileInfoFromIndex(sourceIndex);
    paths.push_back(file->path());
}
// else: it came from the volume half — just skip it, nothing to translate it into
```

But again as I said, approach 2 requires more code definetely than 1. Still wanna give a try to either one of them knowing what would be possile and what not.


> I may succeed in involving you in something better: [lxqt/libfm-qt#417](https://github.com/lxqt/libfm-qt/issues/417)

I can do both ;). Involving in 3,2,1 hahaha.

**Comment by tsujan** [comment]
> bumping Qt to 6.11 would be possible?

Maybe after the next release.

> I'm very persistent 

I too ;) The required code for implementing a feature should be proportional to its value. Making deep changes for a duplicate feature isn't acceptable, unless (1) those changes are minimal and (2) it's 100% guaranteed that they won't have side effects. We don't want to ruin a library for a small feature.

**Comment by tsujan** [comment]
To give you only a simple example of why I'm cautious about deep changes:

I made this small change to circumvent a ["problem"](https://github.com/lxqt/pcmanfm-qt/issues/1816) in GLib: https://github.com/lxqt/libfm-qt/commit/6bdec327f23bc408d8f1512d327ef1fd98c8326b. It seemed like an innocent change, and I didn't think it would result in a side effect, but it did: https://github.com/lxqt/libfm-qt/issues/1085. So, I reversed it in https://github.com/lxqt/libfm-qt/commit/4c47b34d6f9e5cd3645e30925a69a037ab14b12b.
